[Quant] [PT2] Enable QAT Quantization flow in X86InductorQuantizer

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112570
* #111281
* __->__ #111280

**Summary**
This PR enables PT2 QAT Quantization flow in `X86InductorQuantizer`.



**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_with_quantizer_api
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_unary_with_quantizer_api
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

leslie-fang-intel committed:
add ConvBN(ReLU) recipe

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QAT Quantization flow in X86InductorQuantizer"


**Summary**
This PR enables PT2 QAT Quantization flow in `X86InductorQuantizer`.



**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_with_quantizer_api
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_unary_with_quantizer_api
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QAT Quantization flow in X86InductorQuantizer"


**Summary**
This PR enables PT2 QAT Quantization flow in `X86InductorQuantizer`.



**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_with_quantizer_api
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_unary_with_quantizer_api
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168 @andrewor14, Could you kindly help to take a look of this PR? 
One thing to note is per our discussion previously, we prefer small test case since it's easy to debug. So I write separate test case for QAT and PTQ, even some of UTs has same pattern to check as PTQ.

jerryzh168 commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -79,18 +83,26 @@ def clone(x):
-        convert_model = convert_pt2e(prepare_model)
-        torch.ao.quantization.move_exported_model_to_eval(convert_model)
-        return convert_model
+    def _generate_reference_quantized_model(self, mod, inputs, is_qat=False):
```
nit: let's call this qdq_quantized_model, to align with https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_static.html#convert-the-calibrated-model-to-a-quantized-model

leslie-fang-intel replied:
Thanks for the comment and changed.

jerryzh168 commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -79,18 +83,26 @@ def clone(x):
+                else prepare_pt2e(export_model, quantizer)
+            )
+            prepare_model(*inputs)
+            convert_model = convert_pt2e(prepare_model)
```
please add fold_quantize=True here, we are thinking of making this default soon

leslie-fang-intel replied:
Thanks for the comment and add `fold_quantize=True` into `convert_pt2e`.

jerryzh168 commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -626,6 +641,93 @@ def forward(self, x):
+        self._test_common(
+            mod,
+            (v,),
+            4,
+            17,
```
these numbers will be tricky to get right I feel, so ideally this can be structured differently, e.g. like specifying a list of nodes etc.

leslie-fang-intel replied:
Yean, it's tricky to make it correct. I add a following up PR to re-structure it: https://github.com/pytorch/pytorch/pull/112570 in this ghstack. After that, we only need to check the patterns we care in each test case such as: dequant promotion, QConv/Linear Unary and QConv Binary instead of all the patterns . Please help to take a look of that PR also @jerryzh168.

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -398,6 +416,82 @@ def _annotate_for_static_quantization_config(
+                [conv_partition, bn_partition]
+            )
+
+            if (
+                conv_node.op != "call_function"
+                or conv_node.target != torch.ops.aten.conv2d.default
+            ):
```
just making sure, do you have plans to follow https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow later?

leslie-fang-intel replied:
Did you mean use `SubgraphMatcherWithNameNodeMap` to match the pattern for quantizer? I think we can follow up the changes later, maybe after the changes in xnnpack_quantizer. BTW: as discussed in the slack, is there any plan to support  functionalization for pre autograd aten IR? I think it can help us save lots of patterns to use  `SubgraphMatcherWithNameNodeMap` .

jerryzh168 replied:
sg, yeah functionalization will be supported in 1-2 month I think

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -398,6 +416,82 @@ def _annotate_for_static_quantization_config(
+        self, model: torch.fx.GraphModule, config: QuantizationConfig
+    ):
+        # Annotate QAT Specific patterns
+        self._annotate_qat_conv2d_bn_unary(model, config)
```
do you need to add something to prepare_pt2e code to support this pattern? https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantize_pt2e.py#L178

we are also thinking about moving qat specific code to quantizer as well actually, we'll discuss a bit internally and get your feedback

leslie-fang-intel replied:
Yean, for now we only have relu as unary post op which I think is already supported in https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantize_pt2e.py#L178.

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QAT Quantization flow in X86InductorQuantizer"


**Summary**
This PR enables PT2 QAT Quantization flow in `X86InductorQuantizer`.



**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_with_quantizer_api
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_unary_with_quantizer_api
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QAT Quantization flow in X86InductorQuantizer"


**Summary**
This PR enables PT2 QAT Quantization flow in `X86InductorQuantizer`.



**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_with_quantizer_api
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_unary_with_quantizer_api
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168 @andrewor14, thanks for the comments. Please kindly help to take a look again.

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QAT Quantization flow in X86InductorQuantizer"


**Summary**
This PR enables PT2 QAT Quantization flow in `X86InductorQuantizer`.



**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_with_quantizer_api
python -m pytest test_x86inductor_quantizer.py -k test_qat_conv2d_unary_with_quantizer_api
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
